webpack/InsightsCloudSync/InsightsCloudSyncActions.js

```diff
@@ -1,25 +1,49 @@
-import React from 'react';
 import { post } from 'foremanReact/redux/API';
 import { translate as __ } from 'foremanReact/common/I18n';
 import { insightsCloudUrl } from './InsightsCloudSyncHelpers';
-import { INSIGHTS_CLOUD_SYNC } from './InsightsCloudSyncConstants';
-import { foremanUrl } from '../ForemanRhCloudHelpers';
+import {
+  INSIGHTS_CLOUD_SYNC,
+  INSIGHTS_CLOUD_SYNC_TASK,
+} from './InsightsCloudSyncConstants';
+import { setupTaskPolling, taskRelatedToast } from '../common/ForemanTasks';
 
-export const syncInsights = () =>
-  post({
-    key: INSIGHTS_CLOUD_SYNC,
-    url: insightsCloudUrl('tasks'),
-    successToast: response => (
-      <span>
-        {__('Recommendation sync has started: ')}
-        <a
-          target="_blank"
-          rel="noopener noreferrer"
-          href={foremanUrl(`/foreman_tasks/tasks/${response.data?.task?.id}`)}
-        >
-          {__('view the task in progress')}
-        </a>
-      </span>
-    ),
-    errorToast: error => `${__('Recommendation sync has failed: ')} ${error}`,
+export const syncInsights = (fetchInsights, query) => dispatch =>
+  dispatch(
+    post({
+      key: INSIGHTS_CLOUD_SYNC,
+      url: insightsCloudUrl('tasks'),
+      handleSuccess: ({
+        data: {
+          task: { id },
+        },
+      }) => {
+        dispatch(syncInsightsStartedToast(id));
+        dispatch(setupInsightsTaskPolling(id, fetchInsights, query, dispatch));
+      },
+      errorToast: error => syncInsightsError(error),
+    })
+  );
+
+const syncInsightsError = error =>
+  `${__('Recommendation sync has failed: ')} ${error}`;
+
+const syncInsightsStartedToast = taskId =>
+  taskRelatedToast(taskId, 'info', __('Recommendation sync has started: '));
+
+const setupInsightsTaskPolling = (taskId, fetchInsights, query, dispatch) =>
+  setupTaskPolling({
+    taskId,
+    key: INSIGHTS_CLOUD_SYNC_TASK,
+    onTaskSuccess: () => {
+      dispatch(fetchInsights({ query, page: 1 }));
```

Comment:
to fix the double API call, I removed the `dispatch` here:

```suggestion
      fetchInsights({ query, page: 1 });
```

and also reverted the change of return in https://github.com/theforeman/foreman_rh_cloud/pull/561/files#diff-7dfe910e17369a7213356b886dabc440bb547dbec9b42ad7bda757f9b368f3b5R43

but I am not entirely sure why it works to be honest... 
maybe it gets the `dispatch` from the scope above? 🤔

Reply:
I have left the internal return, but removed the dispatch. It was enough to stop the double call.

From looking at the call stacks attached to redux, it looks like first the `fetchInsights` is executed and then the result is passed to `dispatch` which executes it again for some reason.

Inner call:

![Screenshot from 2021-06-13 21-56-55](https://user-images.githubusercontent.com/10046168/121858932-ccd26780-ccff-11eb-888f-e0dffa2f705a.png)

Outer call:

![Screenshot from 2021-06-13 21-55-22](https://user-images.githubusercontent.com/10046168/121858988-dbb91a00-ccff-11eb-956d-aaf54652e2f8.png)

Reply:
awesome! let's do that then :)